GUIDELINES.md

```diff
@@ -0,0 +1,135 @@
+* [Links](#links)
+
+## Goals
+
+* Put a **layer of abstraction** on the system under test: we call these abstractions _steps_ (as derived from Cucumber)
+  * so if the system under test changes, the test system doesn't need to change
+  * so the test just tells _what_ is tested, not _how_ (tests become more business-centric, whenever possible, all the details are in the steps)
+  * so you can reuse test code
+  * to sum up, there is a proper separation of concerns (separation of business rules, workflow and technical implementation) and the ability to reuse test code
+* **Simplify dependency tree creation** by using configuration and automatic dependency resolution
+* That **simplifies test setup** which makes it **cheaper to have bigger integration tests**
+* With that in mind it becomes worth to **invest more in integration tests** instead of unit tests
+  * which **reduces the use of mocks** and
+  * leads to **more realistic tests**
+
+## One-by-one steps
+
+* Create a module testing configuration in the test sources for each module in your project (e.g. for a module `app` create a file `AppModuleTestingConfiguration` with the following code: `val appModuleTestingConfiguration = moduleTestingConfiguration { ... }`; see below for more info on how to add dependencies to the configuration)
+  * Always use the same package structure for test sources, e.g. if you app resides in a module `app` and has a base package `org.myorg.product.app` then put the module test configuration in exactly in the same package
+  * If there is a class `org.myorg.product.app.view.LoginViewModel` then the test class `LoginViewModelTest` and `LoginViewModelSteps` would reside in the exact same package, too
+  * If you have a dedicated module for test sources you should still stick to the same package structure, e.g. if the module is `:app:test` _don't_ add `test` to the package structure for test sources
+  * If modules depend on each other the module testing configuration should do the same, e.g. `val appModuleTestingConfiguration = moduleTestingConfiguration(dependentConfig1, dependentConfig2, ...) { ... }`, this automatically imports all dependent configurations
+* Create a test class for the _feature under test_ (e.g. `LoginTest`), so the test is agnostic of the underlying structures (classes, models, views, data structures, ...)
+  * Should it be that a test really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelTest`)
+* Create a steps class for the feature under test (e.g. `LoginSteps`) which will contain all the `given`, `when` and `then` functions as well as dependency configuration and setup code
+  * The steps class should know about the setup of the test, so for example if the steps class tailored to test the integration of multiple classes, it should also be called something like `LoginIntegrationSteps` which tests a broad stack of classes (e.g. a `LoginViewModel` talking to an `AuthManager` and a `SessionStore`)
```

Comment:
I don't think this naming convention is something which anybody does yet. Maybe get some feedback from more people here.

Reply:
In general I am not 100% sure if we should have multiple steps classes for the same class, just because it is real or mocked or whatever. As long as I can keep it real, I would do so. This is an advantage of sweetest we should also use. As stated in the beginning, we should prefer Steps classes which interact with real objects over faked ones, so I would just reuse them. And when it is really necessary, change the configuration in the steps class above. Not sure about it, but I would rather try to avoid having multiple steps classes and only do this when it is really necessary to have multiple ones.

Reply:
i think that's already 3 questions you're addressing:

1. having separate steps classes for interacting with mock vs. a real implementation (i think it should be seperate)
2. having some kind of naming convention for steps classes (i think it makes sense, we should align on it though)
3. whether how much on an integration level we should test at all (i agree that integration level should be preferred, but that's anyways stated at many places, or at least it will)